File: zemiao/188.py
class Solution(object):
    def maxProfit(self, k, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        delta = [1 if prices[i+1]>prices[i] else 0 for i in range(len(prices)-1)]
        count = 0
        for i in range(len(delta)):
            if delta[i]==1 and (i==0 or delta[i-1]!=1):
                count += 1
        if count < k:
            # fallback to Problem - Stock II
            delta = [prices[i+1]-prices[i] if prices[i+1]>prices[i] else 0 
                     for i in range(len(prices)-1)]
            return sum(delta)
            
        if k == 0:
            return 0
        
        hold = [-32768 for i in range(k)]
        release = [-32768 for i in range(k)]
        for price in prices:
            for i in reversed(range(k)):
                release[i] = max(release[i], hold[i]+price)
                hold[i] = max(hold[i], release[i-1]-price if i>0 else -price)
        
        return max(release + [0])
            

File: tnzhou/188.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        int n = prices.size();
        k = min(n, k);
        vector<int> fbuy(k + 1, -2147483647);
        vector<int> fsell(k + 1, 0);
        for (int i = 0; i < n; i ++) {
            for (int j = 1; j <= k; j ++) {
                fbuy[j] = max(fsell[j - 1] - prices[i], fbuy[j]);
                fsell[j] = max(fbuy[j] + prices[i], fsell[j]);
            }
        }
        int max_profit = 0;
        for (int i = 1; i <= k; i++)
            max_profit = max(max_profit, fsell[i]);
        return max_profit;
    }
};  
